### PR TITLE
Fixing Email Validator to handle edge case

### LINKED ISSUE #18585: EmailValidator fails on valid email address (when IDN is enabled) and address contains .-
### What steps will reproduce the problem?
Validate any (valid) email address with `.-` in the address, and set `enableIDN` to `true`.
i.e. user.-name@domain.com (this is a valid email address).

### What is the expected result?
The email will validate.

### What do you get instead?
The email fails validation.

### Additional info
I traced this problem down. What happens is in framework/validators/EmailValidator.php, line 80
$matches['local'] = $this->idnToAscii($matches['local']);

The `idnToAscii` function returns `false` (because line 142 returns false). It seems that a string with `.-` in it will fail to convert to ascii in this function, causing valid email addresses to fail email validation.

This causes the email address to end up being (in the example above) `@domain.com` when re-assembled on line 81:
$value = $matches['name'] . $matches['open'] . $matches['local'] . '@' . $matches['domain'] . $matches['close'];

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.41.1
| PHP version      |  7.4.16
| Operating system | Linux

### PR DESCRIPTION
**Pull request by ihitbuttons**
Fixing Email Validator to handle edge case where email address is valid but fails ascii conversion.

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #18585

### CODE REVIEW

File: framework/validators/EmailValidator.php [hunk 1]
```diff
@@ -77,7 +77,21 @@ protected function validateValue($value)
             $valid = false;
         } else {
             if ($this->enableIDN) {
-                $matches['local'] = $this->idnToAscii($matches['local']);
+                // https://github.com/yiisoft/yii2/issues/18585
+                $localIDNTest = $this->idnToAscii($matches['local']);
```
Comment (SilverFire) [inline framework/validators/EmailValidator.php:81]: This method now has too many nested conditions.

Could you extract the whole branch to a separate method? Somehow like this:

```suggestion
    $value = $this->validateIdnValue($matches);
```


```php
private function validateIdnValue(&$matches)
{
    $localIDNTest = $this->idnToAscii($matches['local']);

    if ($localIDNTest === false) {
	$newPatternExploded = explode('@', $this->pattern, 2);
	$newPattern = $newPatternExploded[0] . "$/";
	$newFullPatternExploded = explode('@', $this->fullPattern);
	$newFullPattern = $newFullPatternExploded[0] . '@' . $newFullPatternExploded[1]  ."$/";
	if (!preg_match($newPattern, $matches['local']) || ($this->allowName && preg_match($newFullPattern, $matches['local']))) {
	    $matches['local'] = $localIDNTest;
	}
    } else {
	$matches['local'] = $localIDNTest;
    }

    $matches['domain'] = $this->idnToAscii($matches['domain']);
    return $matches['name'] . $matches['open'] . $matches['local'] . '@' . $matches['domain'] . $matches['close'];
}
```

**Review by SilverFire — CHANGES_REQUESTED** [review]
Otherwise good to go

File: framework/validators/EmailValidator.php (hunk 1 above)
Comment (ihitbuttons) [reply framework/validators/EmailValidator.php:81]: Yes, I will refactor it and resubmit it.

Comment (ihitbuttons) [reply framework/validators/EmailValidator.php]: Refactored to keep it more consistent with how it was originally implemented.

**Review by bizley — APPROVED** [review]

File: framework/validators/EmailValidator.php [hunk 2]
```diff
@@ -176,4 +176,28 @@ public function getClientOptions($model, $attribute)
 
         return $options;
     }
+
+    /**
+     * @param string $value
+     * @return string|bool returns string if it is valid and/or can be converted, bool false if it can't be converted and/or is invalid
+     * @see https://github.com/yiisoft/yii2/issues/18585
+     */
+    private function validateIdnValue ($value)
```
Comment (samdark) [inline framework/validators/EmailValidator.php:185]: It is better to separate value validation and value conversion i.e. `$this->idnToAscii($matches['local']);` should stay as is but then it should be validated.

Comment (ihitbuttons) [reply framework/validators/EmailValidator.php:185]: It has already been merged, but I am happy to submit a new pull request with changes.

Not 100% sure I am following you based on where this comment is, but it looks like what you want could be achieved by changing line 80 to:

```
$matches['local'] = $this->idnToAscii($matches['local']); // bring this back
$matches['local'] = $this->validateIdnValue($matches['local']); // this would be the new value for local
```
then removing line 187, as it would be redundant and called outside of this method.

If that is what you are looking for, I can create a new branch and submit it.

Comment (samdark) [reply framework/validators/EmailValidator.php:185]: > It has already been merged, but I am happy to submit a new pull request with changes.

Nope. It was not merged yet. Just push to the same branch and changes will end up in this PR.

Comment (samdark) [reply framework/validators/EmailValidator.php:185]: Checked it again. That's not really validation. Proposed change separately.

File: framework/validators/EmailValidator.php [hunk 3]
```diff
@@ -176,4 +176,28 @@ public function getClientOptions($model, $attribute)
...
+            $newFullPattern = $newFullPatternExploded[0] . '@' . $newFullPatternExploded[1]  ."$/";
+
+            if (preg_match($newPattern, $value) || ($this->allowName && preg_match($newFullPattern, $value))) {
+                return $value;
+            }
+        }
+
+        return $idnTest;
+    }
 }
```
Comment (samdark) [inline framework/validators/EmailValidator.php:203]:
```suggestion
    private function idnToAsciiWithFallback($value)
    {
        $ascii = $this->idnToAscii($value);

        if ($ascii === false) {
            $parts = explode('@', $this->pattern, 2);
            $pattern = $parts[0] . "$/";

            $fullPatternParts = explode('@', $this->fullPattern);
            $fullPattern = $fullPatternParts[0] . '@' . $fullPatternParts[1]  ."$/";

            if (preg_match($pattern, $value) || ($this->allowName && preg_match($fullPattern, $value))) {
                return $value;
            }
        }
        
        return $ascii;
    }
```

Comment (ihitbuttons) [reply framework/validators/EmailValidator.php]: Refactored based on your suggestions

File: framework/validators/EmailValidator.php [hunk 4]
```diff
@@ -176,4 +176,25 @@ public function getClientOptions($model, $attribute)
...
+     * @see https://github.com/yiisoft/yii2/issues/18585
+     */
+    private function idnToAsciiWithFallback($value)
+    {
+        $ascii = $this->idnToAscii($value);
+        if ($ascii === false) {
+            $parts = explode('@', $this->pattern, 2);
+            $pattern = $parts[0] . "$/";
+            $fullPatternParts = explode('@', $this->fullPattern);
+            $fullPattern = $fullPatternParts[0] . '@' . $fullPatternParts[1]  ."$/";
```
Comment (samdark) [inline framework/validators/EmailValidator.php:192]: Why exploding and then imploding?

```suggestion
            $fullPattern = $this->fullPattern ."$/";
```

File: framework/validators/EmailValidator.php (hunk 3 above)
Comment (samdark) [reply framework/validators/EmailValidator.php]: Awesome! I think it got way more readable. One last issue (left a comment) and it's good to go 👍

File: framework/validators/EmailValidator.php (hunk 4 above)
Comment (ihitbuttons) [reply framework/validators/EmailValidator.php:192]: The `pattern` variable contains validation for the full email address [part before the @, the part after the @] when `allowName` is false. 

The `fullPattern` variable is the same, but used when `allowName` is true. 

For this method, we just want to test the first part (the part before the @), not the second part . Using the unaltered regex would cause it to fail.

We could set a new regex, but for functionality reasons, and to hopefully not have to change this check should the regex change, I decided to split the regex and discard the check for the part after the @.

The `fullPattern` has an extra `@` in the beginning:  `/^[^@]...@...$/` vs the `pattern` variable: `/^...@...$/`.

I am by no means a regex expert, but it validates all the test cases like this. When splitting the string by the @ in explode, I needed to re-include it.

The idea being this: The regex used to validate the address works - in some cases, when passed into the ASCII conversion it fails, but ASCII conversion failing does not mean it is an invalid email address, nor does it mean it is a valid email address. So if it fails ASCII conversion in this method (which is just for the first part of the address), we still need to validate it, which we can use part of the regex (but not all of it) that already works to do that.

We could split the regex, as I have done, or tack a known good email address on the back and use the full regex. Or maybe something I didn't see.

If there is another way, open to suggestions. I don't this suggestion would work, however.

Comment (ihitbuttons) [reply framework/validators/EmailValidator.php:192]: The other option would be to take all validation out of this method. If the ASCII conversion results in false, just set it at the original value as it is validated with regex later on.

If that is the case, keeping the ASCII conversion would be done to support backwards compatibility.

Comment (bizley) [reply framework/validators/EmailValidator.php:192]: Since `$this->fullPattern` is subject to change by the configuration the whole check can fail with an error. Isn't it better to provide a new pattern for this only?

Comment (samdark) [reply framework/validators/EmailValidator.php:192]: @bizley assumptions about the pattern value are alright.

Comment (ihitbuttons) [reply framework/validators/EmailValidator.php:192]: The regex pattern could change (I believe I noticed it hasn't changed in several years, but that doesn't preclude it from future change). I put this in place with that in mind, actually. The purpose of this was to not have to maintain two sets of variables (one set being just a subset of another), as this could capture (most of) the changes without having to then understand and/or change a second related variable.

If there are changes in either of these patterns, the same changes would then have to be subsequently reflected in additional private variables.

The pattern would have to fundamentally change the paradigm it is using (i.e. test a set values, require an `@` symbol test another set of values), , not just the content of its check (i.e. the regex for each specific set of values), to break this. At that point, a refactor of several methods, including this one, would be required.

In that respect, it isn't really future proofing yourself or providing any real extra protection, in my opinion (which is why I coded it this way).

If you prefer, I can set these are new private variables and just leave a note that they should be updated whenever the regex to validate the full address is used.

Comment (bizley) [reply framework/validators/EmailValidator.php:192]: It's not about the pattern age here, nothing stops user from setting `->fullPattern = 'lalala'`, since we have these patterns public. We could protect them but it's too late for this now. The solution is to add another pattern so you will not have to operate on the original one, you will only have to use it.

Comment (ihitbuttons) [reply framework/validators/EmailValidator.php:192]: Yeah, that's fine. I will make the changes and re-push the branch.

File: framework/validators/EmailValidator.php [hunk 5]
```diff
@@ -32,6 +32,17 @@ class EmailValidator extends Validator
      * @see allowName
      */
     public $fullPattern = '/^[^@]*<[a-zA-Z0-9!#$%&\'*+\\/=?^_`{|}~-]+(?:\.[a-zA-Z0-9!#$%&\'*+\\/=?^_`{|}~-]+)*@(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?\.)+[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?>$/';
+    /**
+     * @var string the regular expression used to validate the part before the @ symbol, used if ASCII conversion fails to validate the address.
+     * @see http://www.regular-expressions.info/email.html
```
Comment (samdark) [inline framework/validators/EmailValidator.php:37]: `@since 2.0.42`

File: framework/validators/EmailValidator.php [hunk 6]
```diff
@@ -32,6 +32,17 @@ class EmailValidator extends Validator
...
+     */
+    public $patternASCII = '/^[a-zA-Z0-9!#$%&\'*+\\/=?^_`{|}~-]+(?:\.[a-zA-Z0-9!#$%&\'*+\\/=?^_`{|}~-]+)*$/';
+    /**
+     * @var string the regular expression used to validate email addresses with the name part before the @ symbol, used if ASCII conversion fails to validate the address.
+     * This property is used only when [[allowName]] is true.
+     * @see allowName
```
Comment (samdark) [inline framework/validators/EmailValidator.php:43]: `@since 2.0.42`

**Review by samdark — APPROVED** [review]

**Review by machour — APPROVED** [review]

**Review by bizley — APPROVED** [review]

File: framework/validators/EmailValidator.php (hunk 5 above)
Comment (SilverFire) [inline framework/validators/EmailValidator.php:37]:
```suggestion
     * @see http://www.regular-expressions.info/email.html
     * @since 2.0.42
```

File: framework/validators/EmailValidator.php (hunk 6 above)
Comment (SilverFire) [inline framework/validators/EmailValidator.php:43]:
```suggestion
     * @see allowName
     * @since 2.0.42
```

**Review by SilverFire — APPROVED** [review]

**Comment by SilverFire** [comment]
@ihitbuttons Thanks for digging that deep and fixing the bug.

I also appreciate your style of detailed elaboration in the discussions above 👍🏻
